Use permissive wide-to-narrow transcoding in filesystem_error

Chronial opened the pull request:
Fixes #927: `std::filesystem_error` constructor used to raise a `std::system_error` when given paths that are not representable in the current codepage. This replaces unrepresentable characters with a replacement character instead of raising `system_error`.

BillyONeal commented:
This looks reasonable but needs tests.

BillyONeal commented:
I don't know why the CI got unhappy there :(

> This looks reasonable but needs tests.

For example:

BillyONeal commented:
I don't know why the CI is unhappy.

> This looks reasonable but needs tests.

For example:

```
#include <assert.h>
#include <filesystem>
#include <string_view>
#include <system_error>

using namespace std;
namespace fs = std::filesystem;

void test_filesystem_error_with_bad_codepage_characters() {
    fs::path problem_path{L"\U0001F4A3"}; // an emoji that goes boom
    fs::filesystem_error err{"something bad happened", problem_path, error_code{}};
    assert(string_view{err.what()}.find('?') != string_view::npos);
}

int main() { test_filesystem_error_with_bad_codepage_characters(); }
```

Chronial commented:
Thanks, I'll add a test. Is it ok to assume that the tests won't run with a utf-8 system codepage? Not that I have any idea what else to do about that :).

BillyONeal commented:
> Thanks, I'll add a test. Is it ok to assume that the tests won't run with a utf-8 system codepage? Not that I have any idea what else to do about that :).

If I understand correctly getting that requires a manifest we aren't going to be adding to our bits. If you're concerned you can replace my 'emoji that goes boom' with an incorrect surrogate pair or something like that.

Chronial commented:
Added a test. I am not sure whether `WideCharToMultiByte` will always use a `?` as replacement character – according to the docs, that's up to the codepage to specify. I think that's fine, so I rather checked to make sure that the rest of the path is still there.

I also added a check to make sure the test is actually testing something.

Chronial commented:
Regarding the CI failure: This seems unrelated to the contents of this PR.  Running [`python -m pip install --upgrade pip` fails in the setup step](https://dev.azure.com/vclibs/STL/_build/results?buildId=3752&view=logs&j=cdd012bc-41eb-580d-7644-e3ee22f60668&t=1ae3265d-22be-5c49-79f0-b996cadeee7f).

That error is caused by the fact that the python in question is installed in `C:\Program Files` and the command is trying to upgrade a package in there. Apparently the command is executed in a non-elevated context and thus is lacking permission to write to that folder.

I couldn't see what changed compared to the working runs.

StephanTLavavej commented:
We're working on fixing the CI. Thanks for looking into the failures - we're not 100% sure whether the CI should be progressing past the admin-permissions-required upgrade, but clearly that problem needs to be fixed.

cbezault commented:
/azp run

StephanTLavavej commented:
/azp run

StephanTLavavej commented on stl/inc/filesystem:
```diff
@@ -1786,8 +1812,11 @@ namespace filesystem {
             string _Result;
-            const string _Path1_str = _Path1.string();
-            const string _Path2_str = _Path2.string();
+            // Convert the paths to narrow encoding in a way that gracefully handles non-encodeable characters
```
```suggestion
            // Convert the paths to narrow encoding in a way that gracefully handles non-encodable characters
```

StephanTLavavej commented on tests/std/tests/GH_001010_filesystem_error_encoding/test.cpp:
```diff
@@ -0,0 +1,32 @@
+void test_filesystem_error_with_bad_codepage_characters() {
+    fs::path problem_path{L"problematic\xD83Dtestpath"}; // path containing lone high surrogate
+
+    // Make sure the path is not encodeable in our current locale, otherwise the test tests nothing
```
```suggestion
    // Make sure the path is not encodable in our current locale, otherwise the test tests nothing
```

StephanTLavavej commented on tests/std/tests/GH_001010_filesystem_error_encoding/test.cpp:
```diff
@@ -0,0 +1,32 @@
+    }
+    assert(exception_caught);
+
+    // filesystem_error should handle the non-enceable character gracefully when building its message
```
```suggestion
    // filesystem_error should handle the non-encodable character gracefully when building its message
```

StephanTLavavej commented on stl/src/filesystem.cpp:
```diff
@@ -421,6 +421,27 @@ void __stdcall __std_fs_directory_iterator_close(const __std_fs_dir_handle _Hand
+
+    _Result._Err = _Result._Len == 0 ? __std_win_error{GetLastError()} : __std_win_error::_Success;
+
+    // Some Codepages don't support WC_NO_BEST_FIT_CHARS, fall back to default conversion.
```
```suggestion
    // Some codepages don't support WC_NO_BEST_FIT_CHARS, fall back to default conversion.
```

StephanTLavavej commented on tests/std/tests/GH_001010_filesystem_error_encoding/test.cpp:
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <assert.h>
+#include <filesystem>
```
Technically need to include `<exception>` for `std::exception` below:
```suggestion
#include <assert.h>
#include <exception>
#include <filesystem>
```

StephanTLavavej commented on tests/std/tests/GH_001010_filesystem_error_encoding/test.cpp:
```diff
@@ -0,0 +1,32 @@
+    bool exception_caught = false;
+    try {
+        (void) problem_path.string();
+    } catch (std::exception const&) {
```
Nitpicks: This test is already `using namespace std;`. Also, while we previously followed a convention of "every test gets to decide whether it's east `const` or west `const`", we're probably going to change the tests to be consistently west `const` like product code (simpler to explain to new contributors), so we may as well follow that for new tests.
```suggestion
    } catch (const exception&) {
```

StephanTLavavej commented on stl/inc/filesystem:
```diff
@@ -1786,8 +1812,11 @@ namespace filesystem {
-            const string _Path1_str = _Path1.string();
-            const string _Path2_str = _Path2.string();
+            // Convert the paths to narrow encoding in a way that gracefully handles non-encodeable characters
+            const string _Path1_str = _Convert_wide_to_narrow_replace_chars<char_traits<char>>(
+                __std_fs_code_page(), _Path1.native(), allocator<char>());
+            const string _Path2_str = _Convert_wide_to_narrow_replace_chars<char_traits<char>>(
+                __std_fs_code_page(), _Path2.native(), allocator<char>());
```
We've started following a convention of using empty braces to construct temporary objects; this makes code slightly more readable since they don't look like function calls.
Also, while the runtime performance doesn't really matter, there should be a very slight codegen size benefit to extracting the non-inlinable call to `__std_fs_code_page()`. (Note: haven't compiled or tested this, but I've brain-simulated clang-format)
```suggestion
            const auto _Code_page   = __std_fs_code_page();
            const string _Path1_str = _Convert_wide_to_narrow_replace_chars<char_traits<char>>(_Code_page,
                _Path1.native(), allocator<char>{});
            const string _Path2_str = _Convert_wide_to_narrow_replace_chars<char_traits<char>>(_Code_page,
                _Path2.native(), allocator<char>{});
```

Chronial commented:
Thanks for your quick feedback 👍

CaseyCarter commented:
Congratulations on your first STL contribution!
